Dafny Style Guide (#304)
dafny-lang/dafny: Docs/StyleGuide/Style-Guide.md

This style guide provides coding conventions for the Dafny code.

*This documentation is still in progress. Please feel free to add any more suggestions if any.*

## Naming Convention
Any **variables** are named with `camelCase`.

> ```dafny

Author: Nice on the ` ```dafny`, not sure if we have enough leverage to have highlighting support for Dafny added. Maybe we should try? https://github.com/github/linguist

Reviewer: Yeah I think it's a good idea!

var minValue := 1;
var cipherMessage := "Hello World";
```

Any **lemmas**, **predicates**, **functions**, **methods**, **classes**, and **modules** are named with `PascalCase`.
```dafny
method FindIndex(arr: seq<int>, k: int)
    ...
```

Any static or global **constants** are named with `UPPERCASE_WITH_UNDERSCORES`.
```dafny
static const MONTHS_IN_A_YEAR := 12;
```

> 

Author: What about datatypes, datatype constructors and the fields of a datatype? Maybe there are other things I'm not thinking of right now?

Reviewer: For now, I will write `datatype`s and `newtype`s should follow `PascalCase`, but I'm not sure if this is the best practice. What do other people think?

Dafny by default opens a new brace in the same line.
```dafny
module M() {
    ...
    method Met() {
        ...
    }
}
```
In case the method (or function, lemma, etc) signature is too long to fit in one line, or in case the signature has at
least one specification clause, the new brace goes to the new line.

```dafny
module M() {
    ...
    method Met(i: int) returns (j: int)
        requires i % 2 == 0
        ensures j > 10
    {
        ...
    }
}
```

This applies to every scope: `module`, `class`, `predicate`, `if`, `while`, and more.

### Imports
Import the modules with `opened` to avoid typing a full module name each time a method is invoked.
```dafny
import opened StandardLibrary
...
```

When there are more than two modules with the same method name, specify the name of the module during method invocation.

```dafny
import opened MyModule
import opened YourModule
...
method MyMethod() {
    MyModule.foo();
    YourModule.foo();
}
```

In case typing the full module name is cumbersome, import it with a shorthanded name.

```dafny
import M = MyModuleWithACumbersomeName
import Y = YourModuleWithACumbersomeName
...
method MyMethod() {
    M.foo();
    Y.foo();
}

> ```

Reviewer 2: I'd suggest the following for imports:

### Imports

By default, import modules without opening them.

```
import Coffee
...
```

However, if some members of a module are used very frequently, import it using `opened`:

```
import opened Donut
...
```

When a file uses two modules, and both of them define a method of the same name, do not import them `opened`.

```dafny
import MyModule
import YourModule
...
method MyMethod() {
    MyModule.foo();
    YourModule.foo();
}
```

In this case, if you want to shorten their name, import them with a shorthand name.

```dafny
import M = MyModuleWithACumbersomeName
import Y = YourModuleWithACumbersomeName
...
method MyMethod() {
    M.foo();
    Y.foo();
}
```

Common imports, such as `StandardLibrary` and `Native`, should be grouped together, followed by custom module imports
with a blank line in-between.

```dafny
import opened StandardLibrary
import opened Native

import opened Donut
import Coffee

> ```

Author: Should we suggest ordering them (loosely) alphabetically? At least when there are many `import`s/`include`s?

By loosely I mean that it sometimes makes more sense to group `import`s/`include`s logically.

Reviewer: Yeah I agree. I added this suggestion in the guide, but please feel free to modify when needed.

## Indentation and Line Breaks

### Tabs or Spaces?
Spaces are preferred over tabs. Tabs should only be used to remain consistent with the existing code with tabs.

> Use 4 spaces for each indentation.

Reviewer 3: Whoa, why?  I thought all decent languages are 2.  I can only think of Java and Python (neither of which qualify) that use 4.

Author: Rust uses 4. I was a proponent for 2 for a long time, but would vote for 4 for Dafny.

I guess one of the reasons I prefer 4 is that after a `method` (etc.) there are usually more long keywords like `requires`, `ensures`, or `modifies` on the lines right after without an empty line separating them. 2 spaces doesn't give me as much visual separation as 4.

Reviewer 3: (Thanks for Rust example)

I guess it depends on the code you're writing.  I mainly just want to write plain functional programs, and having 4 space indent will be sad for this.  I'll cede the point, but won't promise to follow it when it doesn't make sense, which seems to be almost always in the code we've been writing.

Author: I completely see the point. I can live with either.

Reviewer 4: I like 2. 2 is sufficient for my eyes to clearly notice the indentation, and for deeper levels of indentation, more things stay on my screen with 2.

### Maximum Character Limit

> Although there is no strict requirement, it is generally recommended to have maximum of 120 characters in each line.

Reviewer 2: ```suggestion
Although there is no strict requirement, it is generally recommended to have a maximum of 120 characters per line.
```

### Functions, Methods, Predicates, and Lemmas
Every Dafny method has the following signature.
```dafny
method {:<attributes>} MethodName(param1: Type, param2: Type) returns (ret: Type)
    decreases param1
    requires P()
    modifies param2
    ensures Q()
```

When possible, put MethodName and `returns` statement in the same line, as the keyword `returns` is distinct from other
method specification clauses, such as `modifies`, `requires`, `ensures`, and `decreases`, which can be mixed in any
order (although not recommended for readability), while `returns` should always be right after the parameter

> definitions. Each method specification clause should be in a separate line.

Reviewer 2: I'd prefer this:

When possible, put `MethodName` and `returns` statement on the same line, as the keyword `returns` is distinct from other method specification clauses, such as `decreases`, `requires`, `modifies`,  and `ensures`, which should appear in this order. Each method specification clause should be on a separate line.

Reviewer 3: Minor grammatical point:  put `MethodName` and *the* `returns` statement on the same line.

In case the Method signature is too long, we can break it down.
```dafny
method {:<attributes>} MethodName(param1: Type, param2: Type,
        param3: Type, param4: Type, param5: Type)
    returns (ret1: Type, ret2: Type, ret3: Type, ret4: Type,
        ret5: Type)
    decreases param1
    requires P1()
    requires P2()
    requires P3()
    modifies param2
    modifies param3
    ensures Q1()
    ensures Q2()
```

Multiple `requires` or `ensures` can be combined into one:
```dafny
requires
    P1() &&
    P2() &&
    P3()
```
The same rules apply to `function`, `predicate`, and `lemma` definitions.

## Things to Avoid

> ### Parenthesis

Reviewer 2: ```suggestion
### Parentheses
```

In a lot of cases, Dafny does not require parentheses around expressions. Here are some examples.

* If-Else-While Statements
```dafny
// YES
var i := 1;
while i < 10 {
    ...
    if 1 < i {
        ...
    }
    ...
}

// NO
var i := 1;
while (i < 10) {
    ...
    if (1 < i) {
        ...
    }
    ...
}
```

* Simple Boolean/Arithmetic Expressions
```dafny
// YES
method Collatz(num: nat)
    decreases *

> {

Reviewer 3: Shouldn't this { be on the previous line?

Reviewer 2: I would say no, because I like the following rule (which already is in this guide):
> In case the method (or function, lemma, etc) signature is too long to fit in one line, or in case the signature has at least one specification clause, the new brace goes on a the new line.

Author: I vote for a clear no on this, hard to tell where code starts and the contract ends when not having an empty line there.

Reviewer 3: sg

    var n := num;
    while 1 < n
        decreases *
    { 
        n := if n % 2 == 0 then n / 2 else n * 3 + 1;
    }
}

// NO
method Collatz(num: nat)
    decreases *
{
    var n := num;
    while (1 < n) // unnecessary parentheses
        decreases *
    { 
        n := if ((n % 2) == 0) then (n / 2) else ((n * 3) + 1); // unnecessary parentheses
    }
}
```

> ### Whitespaces

Reviewer 2: Question for native English speakers: Does the plural "whitespaces" exist, or should it always by "whitespace"?

Reviewer 3: Doesn't exist.

Avoid unnecessary whitespaces inside expressions.

#### Type Declaration

> Type declaration should have a form of `variableName: variableType`.

Reviewer 3: *A* type declaration, or Type *declarations* should have the form `variableName: variableType`.

```dafny
// YES
var i: int := 1;
class {:extern} Util {
    var {:extern} Exception: System.String
}

// NO
var i : int := 1; // unnecessary whitespaces
class {:extern} Util {
    var {:extern} Exception : System.String // unnecessary whitespaces
}
```

#### Function, Method, Predicate, and Lemma Declaration
The `function`, `method`, `predicate`, and `lemma` definitions should have a form of
`FunctionName(parameterName: parameterType, ...)`.

```dafny
// YES
function method Foo<int>(i: int): int

// NO
function method Foo<int> (i : int) : int // unnecessary whitespaces
```

Avoid too little or too many whitespaces that reduce the overall readability.

```dafny
// YES
lemma MyLemma<A, B>(x: seq<seq<A>>, y: B) {
    ...
}

// NO
lemma MyLemma <A,B> ( x : seq<seq<A>> , y :B){
    ...
}
```

## Recommendations
This section describes a few recommendations that can help make code more readable and easy to follow, although not
strictly enforced.

### Externs
Try to name them the same in Dafny and C# whenever possible, so that in Dafny we only have to write `{:extern}`, not

> `{:extern "<name>"}`.

Author: Too strict. Dafny doesn't target only C#.

Reviewer: Changed it to "target language".

with a blank line in-between.

```dafny
import opened StandardLibrary
import opened Native

import opened Donut
import Coffee
```

Although not required, it's recommended to keep the order of `import`s and `include`s alphabetically, except when it
makes more sense to group them logically.

## Indentation and Line Breaks

### Tabs or Spaces?
Spaces are preferred over tabs. Tabs should only be used to remain consistent with the existing code with tabs.

Use 4 spaces for each indentation.

### Maximum Character Limit
Although there is no strict requirement, it is generally recommended to have a maximum of 120 characters per line.

### Functions, Methods, Predicates, and Lemmas
Every Dafny method has the following signature.
```dafny
method {:<attributes>} MethodName(param1: Type, param2: Type) returns (ret: Type)
    decreases param1
    requires P()
    modifies param2
    ensures Q()
```

When possible, put MethodName and the `returns` statement on the same line, as the keyword `returns` is distinct from
other method specification clauses, such as `decreases`, `requires`, `modifies`, and `ensures`, which should appear in

> this order. Each method specification clause should be on a separate line.

Reviewer 4: I tend to put the `decreases` clause last, because it's relevant only to mutually recursive functions/methods. In this way, `decreases` feels more like it's more tied to the function/method implementation than it is part of a "public specification".

> 

Reviewer 4: Also say that each specification clause should be indented (as is shown in the example).

In case the Method signature is too long, we can break it down.
```dafny
method {:<attributes>} MethodName(param1: Type, param2: Type,
        param3: Type, param4: Type, param5: Type)
    returns (ret1: Type, ret2: Type, ret3: Type, ret4: Type,
        ret5: Type)
    decreases param1
    requires P1()
    requires P2()
    requires P3()
    modifies param2
    modifies param3
    ensures Q1()
    ensures Q2()
```

Multiple `requires` or `ensures` can be combined into one:
```dafny
requires
    && P1()
    && P2()
    && P3()
```
The same rules apply to `function`, `predicate`, and `lemma` definitions.

## Things to Avoid

### Parentheses

In a lot of cases, Dafny does not require parentheses around expressions. Here are some examples.

* If-Else-While Statements
```dafny
// YES
var i := 1;
while i < 10 {
    ...
    if 1 < i {
        ...
    }
    ...
}

// NO
var i := 1;
while (i < 10) {
    ...
    if (1 < i) {
        ...
    }
    ...
}
```

> 

Reviewer 4: Further examples could show `assert` and `print` statements, whose arguments do not need parentheses (yet I see parentheses used in these places from time to time).

* Simple Boolean/Arithmetic Expressions
```dafny
// YES
method Collatz(num: nat)

>     decreases *

Reviewer 4: I love this example!

{
    var n := num;
    while 1 < n
        decreases *
    { 
        n := if n % 2 == 0 then n / 2 else n * 3 + 1;
    }
}

// NO
method Collatz(num: nat)
    decreases *
{
    var n := num;
    while (1 < n) // unnecessary parentheses
        decreases *
    { 
        n := if ((n % 2) == 0) then (n / 2) else ((n * 3) + 1); // unnecessary parentheses
    }
}
```
### Whitespace

Avoid unnecessary whitespace inside expressions.

#### Type Declaration

> A type declaration should have a form of `variableName: variableType`.

Reviewer 4: Yes, if the type is included at all. It's also okay to leave out the colon and the type, provided Dafny can infer the type.

```dafny
// YES
var i: int := 1;
class {:extern} Util {
    var {:extern} Exception: System.String
}

// NO
var i : int := 1; // unnecessary whitespace
class {:extern} Util {
    var {:extern} Exception : System.String // unnecessary whitespace
}
```

#### Function, Method, Predicate, and Lemma Declaration
The `function`, `method`, `predicate`, and `lemma` definitions should have a form of
`FunctionName(parameterName: parameterType, ...)`.

```dafny
// YES
function method Foo<int>(i: int): int

// NO
function method Foo<int> (i : int) : int // unnecessary whitespace
```

Avoid too little or too many whitespace that reduce the overall readability.

```dafny
// YES
lemma MyLemma<A, B>(x: seq<seq<A>>, y: B) {
    ...
}

// NO
lemma MyLemma <A,B> ( x : seq<seq<A>> , y :B){
    ...
}

> ```

Reviewer 4: These are good examples.

function method Foo<int>(i: int): int

// NO
function method Foo<int> (i : int) : int // unnecessary whitespace
```

> Avoid too little or too many whitespace that reduce the overall readability.

Author: Should be "too much whitespace"
